src/core/transpile.jl

```diff
@@ -2447,3 +2447,52 @@ function transpile(
 
     return circuit
 end
+
+struct RejectGatesOnExcludedPositionsTranspiler <: Transpiler
+    connectivity::AbstractConnectivity
+end
+
+function transpile(
+    transpiler::RejectGatesOnExcludedPositionsTranspiler,
+    circuit::QuantumCircuit,
+)::QuantumCircuit
+
+    (found_invalid_gates, message) =
+        are_gates_at_excluded_positions(transpiler.connectivity, circuit)
+
+    if found_invalid_gates
+        throw(DomainError(transpiler.connectivity, message))
+    end
+
+    return circuit
+end
+
+function are_gates_at_excluded_positions(
+    connectivity::AllToAllConnectivity,
+    circuit::QuantumCircuit,
+)::Tuple{Bool,String}
+
+    return (false, "")
+end
+
+function are_gates_at_excluded_positions(
+    connectivity::Union{LineConnectivity,LatticeConnectivity},
+    circuit::QuantumCircuit,
+)::Tuple{Bool,String}
+
+    excluded_positions = get_excluded_positions(connectivity)
+    for instruction in get_circuit_instructions(circuit)
+        qubits = get_connected_qubits(instruction)
+        for (qubit, position) in Iterators.product(qubits, excluded_positions)
```

Comment:
Minor, but for readability, let's change this to be (untested)

```julia
for qubit in qubits
  if qubit in excluded_positions
```

The above style is common in Python where the interpreter consumes CPU cycles in explicit for loops, in Julia there is usually no benefit to reducing nested levels of for loops.